과제 1 - Counter 앱 만들고 파일 분리하기 제출합니다.

Heejae0704 opened the pull request:
기능을 구현하고 테스트 및 ESlint 확인하였습니다.

Heejae0704 committed:
Create counter app

목표 기능을 구현하였습니다.
모든 테스트를 통과하였으며 ESLint에 문제가 없었습니다.

Heejae0704 committed:
Event handler can process event object

Button 컴포넌트에 더해야 할 숫자를 주지 않고 함수 패러미터의 기본값을 쓸 수 있지 않을까
생각해서 () => onClick()이 아니라 onClick 함수 자체를 전달해보니, 더해야 할 숫자
기본값 대신 Object가 전달되었습니다. 아마도 event object가 암묵적으로 전달되는 것으로
이해하고, handleClick 함수에 event object를 첫번째 패러미터로 명시하였습니다.

지금은 쓸 일이 없지만 향후에 만약 이벤트 오브젝트를 사용할 일이 있다면 이 함수를 별 문제없이
사용할 수 있도록 만들어두는 것이 좋지 않을까 해서 이렇게 변경하였습니다.

Heejae0704 committed:
State explicitly refer to previous State

React 문서를 참고하여 count가 이전 count를 참고하여 새로운 count를 만드는 경우이므로
useState가 명시적으로 prevState를 인수로 받아 숫자를 더해주는 로직으로 수정하였습니다.

KJoobin commented on src/components/App.jsx:
```diff
@@ -0,0 +1,25 @@
+function App() {
+  const [state, setState] = useState({
+    count: 0,
+  });
```
리액트에서 state 를 쓰는 방법이 2가지가 있는데 알고계신가요?!

희재님께서 1번 문법을 쓰신 이유가 궁금합니다!! 설명해주실수 있을까요?!
```
//1
const [state, setState] = useState({
  count: 0,
})

//2
const [count, setCount] = useState(0)
```

Heejae0704 replied:
사실 2번을 알고 있었는데 이번에 1번 방법을 사용하면 prevState를 이용하여 state object의 특정 property 과거 값을 다른 다른 property에 사용할 수 있다는 점이 편리해보여서 여기서도 연습삼아 써봤습니다. ^^; 사실은 assignment 2에서 좀 더 의미가 있었을 거 같고, 이 과제에 대해서는 1번을 사용할 메리트는 없는 거 같습니다. 해당 내용은 2번처럼 수정하여 commit하였습니다.

KJoobin commented on src/components/Page.jsx:
```diff
@@ -0,0 +1,19 @@
+import Button from './Button';
+import NumberButtons from './NumberButtons';
+
+function Page({ count, onClick }) {
+  return (
+    <>
+      <p>Counter</p>
+      <p>
+        <Button count={count} onClick={onClick} />
+      </p>
+      <p>
+        <NumberButtons onClick={onClick} />
+      </p>
+    </>
+  );
+}
```
`App.jsx` 가 아닌,  `Page.jsx` 를 만드셔서 작업하신 이유가 있으실까요?!

Heejae0704 replied:
App에서는 state의 정의와 handleClick 등 state를 변화시키는 로직 등 비즈니스 로직을 처리하고, Page는 화면을 그려주는 역할을 하는 구조로 작업하였습니다. 강의에서 나왔던 관심사 분리를 연습하였습니다.

KJoobin commented on src/components/NumberButtons.jsx:
```diff
@@ -0,0 +1,12 @@
+import React from 'react';
+import NumberButton from './NumberButton';
+
+function NumberButtons({ onClick }) {
+  return (
+    <>
+      {[1, 2, 3, 4, 5].map((i) => <NumberButton key={i} number={i} onClick={onClick} />)}
+    </>
+  );
+}
```
NumberButtons 과, NumberButton 을 만드신 이유가 있으실까요?!

Heejae0704 replied:
버튼 자체를 그려주는 역할과 버튼을 모아서 보여주는 역할이 분리될 수 있다고 생각해서 별도로 만들었습니다. 나중에 버튼의 모양이나 구성을 바꿀 때에는 NumberButton에서만 작업을 하면 될 것이고, 버튼의 숫자가 달라지거나 할 때에는 NumberButtons에서 변경을 하면 될 것 같습니다.

KJoobin replied:
`NumberButton` 의 경우에는 희재님의 의견에 동의합니다!! 하지만 `NumberButtons` 의 경우, 따로 분리하는게 오히려 유지보수 측면에서 관리가 힘들거같다는 생각이 들어요!! (소프트웨어 에서는 정답은 없으니깐, 제 의견을 듣고 희재님의 생각을 편하게 말씀해주시면 됩니다!)

A, B, C 페이지에서 `NumberButtons` 를 사용하게 되는데, A 에서는 1~5 까지만 보여주고, B 에서는 1 ~ 10, C 에서는 5 ~ 8 까지만 보여주고 싶을때, 해당 컴포넌트는 재사용 하기가 어려워보여요..!!

만약 `Number` 처럼 제네릭 한것이 아닌경우에는 따로 분리하는것도 의미가 있어보여요!

Heejae0704 replied:
엇 맞습니다. `NumberButtons`는 분리할 필요가 없어보여요. `NumberButton`을 왜 `NumberButtons`와 분리했는지만 질문하신 줄 알고 그것만 생각하고 있었네요. 지금보니 `Page`도 지나치게 쪼갠 것으로 보여서 수정하는 김에 `App`으로 통합하였습니다.

KJoobin commented on src/components/NumberButton.jsx:
```diff
@@ -0,0 +1,9 @@
+import React from 'react';
+
+function NumberButton({ key, number, onClick }) {
+  return (
+    <button type="button" key={key} onClick={(e) => onClick(e, number)}>{number}</button>
+  );
+}
```
넘기신 props 값이 다 정상적으로 나오는지 체크해보셨나요!?

Heejae0704 replied:
React Developer Tool로 확인해보니 key는 prop으로 전달되지 않네요. 해당 내용 제외하고 수정하여 commit 하였습니다.

KJoobin reviewed (commented):
안녕하세요 희재님! 저는 8주간 희재님의 코드를 리뷰할 김주빈 입니다 :)

커밋 메세지를 보니 기능구현에서 그치지 않고, 리팩토링 까지 진행해 주신거같아요!! 👍 👍 
커밋 메세지를 세세하게 남겨주셔서 희재님이 어떤걸 고민하셨고, 어떤 방법을 도출하였는지 알수있어서 보기 좋은거같습니다!

>지금은 쓸 일이 없지만 향후에 만약 이벤트 오브젝트를 사용할 일이 있다면 이 함수를 별 문제없이
사용할 수 있도록 만들어두는 것이 좋지 않을까 해서 이렇게 변경하였습니다.

저는 개인적으로 지금 당장 쓰지 않을 매개변수를 받는걸 좋아하지않아요!!  그 이유는 해당 함수의 이름과 매개변수를 보고 해당 함수가 어떻게 사용될건지 예측을 하게 되는데, 당장 쓰지 않을 매개변수를 미리 작성해 놓으면 방해가 된다고 생각하기 때문이에요...

> React 문서를 참고하여

공식문서를 참고해서 작업을 해주신건 정말 잘해주신거 같습니다..!! 기술을 익힐때 공식문서만한곳이 없는거같아요!! 

컴포넌트를 만들때 고려하는게 어떤것인가요?? 희재님의 생각을 편하게 공유해주세요!!

Heejae0704 committed:
Refactor handleClick function

사용하지 않는 event object를 패러미터에서 제외하여 함수의 사용법을 좀 더 명확하게
드러냈습니다. 그에 따라 event가 자동으로 전달되지 않도록 이 함수가 사용되는 위치에서
화살표 함수를 사용하였습니다.

Heejae0704 commented:
> 저는 개인적으로 지금 당장 쓰지 않을 매개변수를 받는걸 좋아하지않아요!! 그 이유는 해당 함수의 이름과 매개변수를 보고 해당 함수가 어떻게 사용될건지 예측을 하게 되는데, 당장 쓰지 않을 매개변수를 미리 작성해 놓으면 방해가 된다고 생각하기 때문이에요...

저도 동의합니다. 매개변수를 굳이 맞춰주어야 할 상황도 아닌데 굳이 이벤트 객체가 매개변수로 필요하진 않겠네요.

> 컴포넌트를 만들때 고려하는게 어떤것인가요?? 희재님의 생각을 편하게 공유해주세요!!

화면(사용자 UI)에서 독립적인 역할을 할 수 있는 단위로 쪼갤 수 있는 요소가 된다면 별도의 컴포넌트로 만들 수 있다고 생각합니다.
강의에서 들었던 것처럼 비즈니스 로직의 처리와 로직에 따라 화면을 그려주는 역할도 분리해서 각각 컴포넌트로 만들었고요.
그리고 여러 곳에서 동일한 형태로 재사용이 가능한 덩어리가 있다면 이것도 컴포넌트로 만들어 주면 좋다고 생각합니다.

Heejae0704 committed:
Refactor how to handle counter as state

state object 대신에 counter 변수 자체를 state로 관리하는 방식으로 변경했습니다.
관리해야 할 상태가 counter 단 하나라서 prevState의 다른 property들을 참조할 필요도
없기 때문에 가급적 단순하게 만드는 것이 좋겠다고 생각했습니다.

Heejae0704 committed:
Revise props passed to NumberButton

key는 list에서 각 component의 고유성을 나타내는 용도로 사용되며 component에 prop으로
전달되지는 않으므로 NumberButton의 prop 설정에서 제외하였습니다.

KJoobin commented on src/components/App.jsx:
```diff
@@ -0,0 +1,19 @@
+  const [count, setCount] = useState(0);
+
+  function handleClick(numberToBeAdded = 1) {
+    setCount(count + numberToBeAdded);
```
여기에서도, 기존에 사용하시던 `prevState` 값을 이용하셔서 업데이트 하시는건 어떠신가요?!

Heejae0704 replied:
아 맞네요. state가 object일때만 쓸 수 있는 건 아닌데 왜 그 생각을 못했을까요!! 수정했습니다.

KJoobin commented on src/components/Button.jsx:
```diff
@@ -0,0 +1,9 @@
+
+function Button({ count, onClick }) {
+  return (
+    <button type="button" onClick={() => onClick()}>{`Click me(${count})`}</button>
```
```suggestion
    <button type="button" onClick={onClick}>
      {`Click me(${count})`}
    </button>
```
이렇게 작성하시는게 가독성이 더 좋은거 같은데 어떠신가요?!

Heejae0704 replied:
네 훨씬 보기가 좋은 거 같습니다!

KJoobin reviewed (commented):
잘 진행해주셨습니다!!

`Button` 과 `NumberButton` 을 한개로 통일과 지금처럼 분리를 하게되면 각각의 장단점에 대해서 같이 생각해볼까요 ?!

Heejae0704 commented:
> `Button` 과 `NumberButton` 을 한개로 통일과 지금처럼 분리를 하게되면 각각의 장단점에 대해서 같이 생각해볼까요 ?!

만약에 통일을 하게 된다면... 
1. 일단 전달되어야 할 패러미터도 버튼에 표시할 내용, 카운터, 더할 숫자, 그리고 click handler 네 가지나 되어야 할 거 같고요.. 물론 안쓰는 것들은 전달하지 않으면 되긴 하겠지만...
2. 그리고 버튼을 그릴 때, 표시할 내용이 없다면 더할 숫자를 보여주고, 표시할 내용이 있다면 표시해야 하는 로직이 들어가서, guard clause를 이용한 로직이 추가가 될 거 같습니다.

장점은 하나의 컴포넌트를 재사용 가능하게 된다는 점이 있겠고요. 그래서 버튼의 종류가 3가지 이상이 된다면 하나의 컴포넌트를 재사용하는 패턴을 만드는 것도 좋겠다는 생각이 듭니다. 지금은 딱 두 가지라서, 하나는 변화하는 state를 보여주는 특성이 있고, 다른 하나는 그냥 숫자 input을 위한 버튼이라 둘로 나누어 놓아도 되지 않을까 생각해요.

Heejae0704 committed:
Refactor further to reduce components

1. 불필요한 Components 쪼개기를 지양하고 적절한 수준의 분리를 생각하였습니다.
- 그 결과 Page와 NumberButtons는 지나친 분리라고 생각하여 다시 통합하였습니다.

2. 그 밖의 리뷰 코멘트를 반영하고 코드 품질을 올렸습니다.
- 좀 더 이해하기 쉽고, 보기 좋은 코드가 되도록 수정하였습니다.

KJoobin commented:
> 장점은 하나의 컴포넌트를 재사용 가능하게 된다는 점이 있겠고요. 그래서 버튼의 종류가 3가지 이상이 된다면 하나의 컴포넌트를 재사용하는 패턴을 만드는 것도 좋겠다는 생각이 듭니다. 지금은 딱 두 가지라서, 하나는 변화하는 state를 보여주는 특성이 있고, 다른 하나는 그냥 숫자 input을 위한 버튼이라 둘로 나누어 놓아도 되지 않을까 생각해요.

맞는거같아요! 충분히 설득되었습니다!! ㅎㅎ 두개로 나눠서 진행하죠 :)

KJoobin commented on src/components/App.jsx:
```diff
@@ -0,0 +1,27 @@
+import Button from './Button';
+import NumberButton from './NumberButton';
+
+function App() {
```
추후에 page 가 2개 이상일때는 각각의 페이지가 App 에서 추가될거지만 지금은 1개이기 때문에 App 에서 진행을 해도 괜찮을거같아요 :)

KJoobin commented on src/components/Button.jsx:
```diff
@@ -0,0 +1,11 @@
+
+function Button({ count, onClick }) {
+  return (
+    <button type="button" onClick={() => onClick()}>
```
```suggestion
    <button type="button" onClick={onClick}>
```

이렇게 쓰는건 어떨까요?!

Heejae0704 replied:
제안해주신 것처럼 수정하면 함수의 첫 번째 인자로 원래 의도했던 numberToBeAdded 패러미터에 원래 default로 설정했던 값인 1이 들어가는 것이 아니라 SyntheticBaseEvent가 전달되어 숫자가 더해지지 않고 [object object]가 표시되는 버그가 나옵니다. 

핸들러 함수에 guard clause를 넣어서 이벤트가 전달되면 별도 로직으로 동작하게끔 할까도 고민했습니다만, 항상 증가해야 할 숫자를 명시적인 인자로 넣어주는 편이 좋을 거 같습니다. 그래서 코드는 `onClick={() => onClick(1)}`로 간단히 수정했습니다. (핸들러 함수에서는 디폴트 값 설정을 빼주었습니다.)

KJoobin reviewed (approved):
잘 진행해주셨습니다 :)

회고 올라온 뒤 머지하겠습니다!!

Heejae0704 committed:
Changed onClick event handler call

- 1을 더할 때에는 명시적으로 이벤트 핸들러에 명시적으로 1 전달
